layers/+tools/lsp/packages.el

```diff
@@ -0,0 +1,64 @@
+(defconst lsp-packages
+  '(
+    company-lsp
+    (helm-xref :requires helm)
+    (ivy-xref :requires ivy)
+    lsp-mode
+    lsp-ui
+    markdown-mode  ;; depended by lsp-ui
+    ))
+
+(defun lsp/init-company-lsp ()
+  (use-package company-lsp
+    :after company
+    :init
+    ;; Language servers have better idea filtering and sorting,
+    ;; don't filter results on the client side.
+    (setq company-transformers nil
+          company-lsp-async t
+          company-lsp-cache-candidates nil)
+    ;; (spacemacs|add-company-backends :backends company-lsp :modes c-mode-common)
+    ))
+
+(defun lsp/init-helm-xref ()
+  (use-package helm-xref
+    :config
+    ;; This is required to make xref-find-references not give a prompt.
+    ;; xref-find-references asks the identifier (which has no text property) and then passes it to lsp-mode, which requires the text property at point to locate the references.
+    ;; https://debbugs.gnu.org/cgi/bugreport.cgi?bug=29619
+    (setq xref-prompt-for-identifier
+          '(not xref-find-definitions xref-find-definitions-other-window xref-find-definitions-other-frame xref-find-references spacemacs/jump-to-definition))
+
+    ;; Use helm-xref to display xref.el results.
+    (setq xref-show-xrefs-function 'helm-xref-show-xrefs)
+    ))
+
+(defun lsp/init-ivy-xref ()
+  (use-package ivy-xref
+    :config
+    (setq xref-prompt-for-identifier
+          '(not xref-find-definitions xref-find-definitions-other-window xref-find-definitions-other-frame xref-find-references spacemacs/jump-to-definition))
+
+    ;; Use ivy-xref to display xref.el results.
+    (setq xref-show-xrefs-function 'ivy-xref-show-xrefs)
+    ))
+
+(defun lsp/init-lsp-mode ()
+  (use-package lsp-mode
+    :config
+    (require 'lsp-imenu)
+    (add-hook 'lsp-after-open-hook #'lsp-enable-imenu)
+    ;; Disable lsp-flycheck.el in favor of lsp-ui-flycheck.el
+    (setq lsp-enable-flycheck nil)
+    ))
+
+(defun lsp/init-lsp-ui ()
+  (use-package lsp-ui
+    :after lsp-mode
+    :after markdown-mode
+    :config
+    (add-hook 'lsp-mode-hook #'lsp-ui-mode)
+    ))
+
+(defun lsp/init-markdown-mode ()
+  (use-package markdown-mode))
```

Comment:
You don't need to  add markdown to this layer, `package.el` will install it implicitly for us.

On the `init` functions:
There can be only one `init-<package>` function, the layer defining this function is told to be to owner of `<package>`. For instance the `markdown` layer defines the function `markdown/init-markdown-mode` and it makes this layer the owner of the `markdown-mode` package. 
Other layers that extend the configuration of `<package>` define function names `pre-init-<package>` and/or `post-init-<package>` which are respectively executed before and after the `init-<package>` function.
A package is installed and configured only if its `init-<package>` functions is declared which means that the user added the layer owning this package in their dotfile, we say that the layer/package is used when it is explicitly listed in the dotfile.
Therefore if the layer owning a package is not used (not in the dotfile) then the `init-<package>` function is not declared and then `pre-init` and `post-init` functions are not called.
This is the basic principle in Spacemacs to achieve isolation of configuration.

Now in this particular case, even if the `markdown-mode` package is not used (i.e. the user has not added the `markdown` layer to the dotfile) `package.el` will still install the dependencies for `lsp-ui` so `markdown-mode` will be installed, but not configured by Spacemacs.

Reply:
As a side note, a layer created by a user can redefine the `init-<package` and add this layer at the end of the layer list in the dotfile, we say that the layer steals the ownership of `<package>` as the last layer in the list to declare the `init-<package>` function wins the ownership.
This mechanism allows users to be able to reconfigure all pieces of configuration provided by Spacemacs.

More info in https://github.com/syl20bnr/spacemacs/blob/develop/doc/LAYERS.org

Reply:
Thx. Removed `lsp/init-markdown-mode`... Yeah, I do not know much about spacemacs that was why I didn't create the layer earlier.